language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/ExtractTypeCodeAction.java

```diff
@@ -0,0 +1,118 @@
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * The code action to extract anonymous records into type defs.
+ *
+ * @since 2201.0.x
+ */
+@JavaSPIService("org.ballerinalang.langserver.commons.codeaction.spi.LSCodeActionProvider")
+public class ExtractTypeCodeAction extends AbstractCodeActionProvider {
+
+    private static final String RECORD_NAME_PREFIX = "Record";
+
+    public ExtractTypeCodeAction() {
+        super(List.of(CodeActionNodeType.RECORD));
+    }
+
+    @Override
+    public List<CodeAction> getNodeBasedCodeActions(CodeActionContext context, NodeBasedPositionDetails posDetails) {
+        if (posDetails.matchedTopLevelNode().kind() != SyntaxKind.RECORD_TYPE_DESC) {
+            return Collections.emptyList();
+        }
+
+        RecordTypeDescriptorNode node = (RecordTypeDescriptorNode) posDetails.matchedTopLevelNode();
+        if (node.parent().kind() == SyntaxKind.TYPE_DEFINITION) {
+            return Collections.emptyList();
+        }
+
+        Node tlNode = node;
+        while (tlNode.parent().kind() != SyntaxKind.MODULE_PART) {
+            tlNode = tlNode.parent();
+        }
+
+        Set<String> visibleSymbolNames = context.visibleSymbols(context.cursorPosition()).stream()
+                .map(Symbol::getName)
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .collect(Collectors.toSet());
+        String typeName = generateTypeName(RECORD_NAME_PREFIX, visibleSymbolNames);
+
+        LinePosition tlNodeStartLine = tlNode.lineRange().startLine();
+        Range extractedRecordRange = new Range(CommonUtil.toPosition(tlNodeStartLine),
+                CommonUtil.toPosition(tlNodeStartLine));
+        Range originalNodeRange = new Range(CommonUtil.toPosition(node.lineRange().startLine()),
+                CommonUtil.toPosition(node.lineRange().endLine()));
+
+        String typeDesc = String.format("type %s %s;%n%n", typeName, node.toSourceCode());
+        try {
+            typeDesc = Formatter.format(typeDesc);
+        } catch (FormatterException e) {
+            LSClientLogger.getInstance(context.languageServercontext()).logError(LSContextOperation.TXT_CODE_ACTION,
+                    "Failed to format extracted source: " + typeDesc, e, null, (Position) null);
+            return Collections.emptyList();
+        }
+        TextEdit typeDescEdit = new TextEdit(extractedRecordRange, typeDesc);
+        TextEdit replaceEdit = new TextEdit(originalNodeRange, typeName);
+        CodeAction codeAction = createCodeAction(CommandConstants.EXTRACT_TYPE, List.of(typeDescEdit, replaceEdit),
+                context.fileUri(), CodeActionKind.RefactorExtract);
+        return List.of(codeAction);
+    }
+
+    @Override
+    public String getName() {
+        return "ExtractType";
+    }
+
+    private static String generateTypeName(String prefix, Set<String> visibleSymbolNames) {
```

Comment:
Shall we add this method  or reuse one of the methods in CommonUtil to generate the type name
https://github.com/ballerina-platform/ballerina-lang/blob/02411da0cf01ac4f76c768ed911728b0e0ad115b/language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/utils/CommonUtil.java#L718